### ISSUE TITLE
[feature request] Ability to customize getState

### ISSUE THREAD

**Issue by Andarist** [reactjs/react-redux#870]
I'm writing a custom abstraction over a redux's store, something similar to https://github.com/HenrikJoreteg/redux-bundler .

I wouldn't like to diverge from redux's core too much (which includes react-redux ofc) but I bind selectors and actions to the **store itself**. While I understand this is not "idiomatic redux", this is somewhat crucial for us - we'd like to have ability to extend store at runtime & pass a single API object to custom extensions so can consumer/use it easier without worrying about using `dispatch`/`getState` appropriately.

Because of this it only makes sense for us to use existing bound actions/selectors (which exist on the store) in our `mapStateToProps` etc. This is not possible at the moment, because the only thing we can get there is `store.getState()` result.

So ideally we'd like to have a way to access store directly, I've explored bunch of things - also creating our own connect from scratch on top of the connectAdvanced API but actually getState call is hardcoded in the connectAdvanced implementation.

It seems that the easiest way would be to extend allowed options with something like `createGetState` which we could use [here](https://github.com/reactjs/react-redux/blob/76dd7faa90981dd2f9efa76f3e2f26ecf2c12cf7/src/components/connectAdvanced.js#L122):
```js
this.store = props[storeKey] || context[storeKey]
this.getState = typeof createGetState === 'function' ? createGetState(this.store) : store.getState
```

and later just pass it appropriately to [this place](https://github.com/reactjs/react-redux/blob/76dd7faa90981dd2f9efa76f3e2f26ecf2c12cf7/src/components/connectAdvanced.js#L16) and refactor the code to
```diff
-function makeSelectorStateful(sourceSelector, store) {
+function makeSelectorStateful(sourceSelector, getState) {
  // wrap the selector in an object that tracks its results between runs.
  const selector = {
    run: function runComponentSelector(props) {
      try {
-        const nextProps = sourceSelector(store.getState(), props)
+        const nextProps = sourceSelector(getState(), props)
        if (nextProps !== selector.props || selector.error) {
          selector.shouldComponentUpdate = true
          selector.props = nextProps
          selector.error = null
        }
      } catch (error) {
        selector.shouldComponentUpdate = true
        selector.error = error
      }
    }
  }

  return selector
}
```

I'm ofc willing to work on this once this gets approval and we establish the API.

**Comment by timdorr** [comment]
It sounds like what you're describing (and what redux-bundler looks like) is "Redux" in name only. 

In redux-bundler, there is [extensive enhancement going on](https://github.com/HenrikJoreteg/redux-bundler/blob/master/src/compose/index.js), which adds new APIs and many behavioral changes. I would argue that it's not really "Redux" at that point. It looks and sounds similar, but it's really a new thing.

Basically, if you don't implement getState() the same way, then you're probably not Redux. And you should probably not be using React Redux bindings.

**Comment by Andarist** [comment]
My "custom" implementation is way closer to redux than `redux-bundler`, I've used the latter as point of reference as mine is not OSS at this point.

I'm using regular redux machinery under the hood, only with **thin** layer of additional abstraction on top of it - in form of prebound actions and selectors attached to the store itself.

It's simply something along those lines:
```
const store = createStore(
	createReducer(normalizeState(preloadedState)),
	undefined,
	compose(applyMiddleware(...middlewares)))
)

const finalStore = assign(store, bindActionMethods(actions, store), bindSelectors(selectors, store.getState))
```

Basically its just a sugar over vanilla redux's store, nothing more. I can technically reuse my selectors in `mapStateToProps` but it's a lot more of boilerplate that I could avoid if only I had direct access to the store there.

**Comment by markerikson** [comment]
I haven't yet had a chance to dig into `redux-bundler`'s implementation yet, but in general, the "right" way to expand the store's capabilities or change the behavior of the standard store API is by writing a store enhancer that returns its own version of `getState` or `subscribe` rather than the built-in version.  See my examples here:

http://blog.isquaredsoftware.com/presentations/2017-09-might-need-redux-ecosystem/#/26

**Comment by markerikson** [comment]
@Andarist : maybe I'm misunderstanding your question and use case, but it really does seem like a store enhancer is the right answer here.  You can provide your own version of `getState`, which can return whatever it wants instead of / in addition to the original store's return value.  Here's an example:

```js
export const meaningOfLifeEnhancer = (createStore) => {
    return (rootReducer, preloadedState, enhancers) => {
        const store = createStore(reducer, preloadedState, enhancers);
        
        function newGetState() {
            return {
                ...store.getState(),
                meaningOfLife : 42
            };
        }
        
        return { ...store, getState : newGetState };
    }
}


const composedEnhancers = compose(
    middlewareEnhancer, 
    meaningOfLifeEnhancer,
);

const preloadedState = {counter : 2};

const store = createStore(rootReducer, preloadedState, composedEnhancers);
console.log(store.getState());
// {counter : 2, meaningOfLife : 42}
```

**Comment by Andarist** [comment]
With the proposed approach I won't be able to access `otherFunction` from `react-redux` bindings.

The best hacky solution for now that I came up with is smth like this:
```js
const proxyStore = Object.create(store)
// react-redux binding should call `getState` under the hood and pass its result as argument to mapStateToProps
proxyStore.getState = () => store
// ...
<Provider store={proxyStore}>
```
**note** - i haven't yet tested it

So I could later do
```js
const mapStoreToProps = store => ({
  value: store.getValue(),
})
export default connect(mapStoreToProps)(MyComponent)
```

And the last given snippet is somewhat what I aim for.

**Comment by markerikson** [comment]
@Andarist : can you show me exactly where and how you need to access the store object itself, rather than just the state reference?

**Comment by Andarist** [comment]
I'd like to be able to do (without a hack) exactly what the last snippet shows.

We are building an extensible system with plugins written by 3rd party authors and we need to limit what they can do (yeah, ofc part of that could be done with a middleware and scoped dispatch) with more friendly API than bare `getState`/`dispatch`.

**Comment by markerikson** [comment]
It seems you've got two main options:

- A store enhancer that overrides `getState` to return something other than just the plain store state
- Your own alternative to `connect` that behaves differently.

The current API is just not intended to expose the _store_ object itself to "userland" code - it's meant to act as an abstraction layer that hides the store entirely.

**Comment by timdorr** [comment]
The ultimate point here is that this is a really specialized edge case request. You've got your options laid out, but it's not something we're going to add to the API right now.
